pennylane/registers.py

```diff
@@ -0,0 +1,88 @@
+# Copyright 2018-2024 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This module contains the :func:`registers` function.
+"""
+
+from .wires import Wires
+
+
+def registers(register_dict):
+    """Returns registers of wire registers as a dictionary of register names and
+    :class:`~pennylane.wires.Wires`.
+
+    The registers are a dictionary of :class:`~.Wires` objects where the key is the register name
+    and the value is the ``Wires`` object. The values can also be another registers object. This
+    allows creating a nested dictionary of registers.
+
+    Args:
+        register_dict (dict): a dictionary where keys are register names and values are either
```

Comment:
If you tried that you would get a `ValueError` since the function only takes either a `dict` or `int`. The PRD does not specify allowing a `Wires` instance, maybe we should though.

Reply:
@obliviateandsurrender we did think about this when devising a UI, but we came to these conclusions:

1. (a) `Wires` instances being accepted poses a question for how _other_ wires objects should be ordered. E.g., how do we expect to index / label wires in this example?

```
qml.registers({"reg1": 3, "reg2": qml.wires.Wires(["A", 1, 42]), ...})
```

(b) It would also require an additional search through the entire dictionary, an additional complexity overhead that, yes, may be peanuts now 🥜, but might not be with millions of qubits 📈! 

2. If you require specific qubit indices to be mapped onto specific physical qubits (e.g., for hardware constraints and "bad" qubits), we recommend just sticking with `qml.wires.Wires` and providing customized labels _or_ you can use `qml.registers` and map your register's wire labels to other desired labels with `Wires.map`.

3. You can use an already-instantiated `Wires` object with `qml.registers`, but you should be careful because of 1(a) 😅. You can do this:

```
wires_dict = qml.registers({"reg1": 3, "reg2": 2})
wires_dict["new_reg"] = already_instantiated_wires
```

But `already_instantiated_wires` might have overlapping wire labels (might be unintentional!)

Reply:
Your concern is valid to have, but I think the feature is a great starting point. If there's good user signal to make some additions to `registers` that relate to your concern, we'll look into how we can make it better!